src/adplug/adplug-xmms.cc

```diff
@@ -291,14 +328,139 @@ const char * const AdPlugXMMS::defaults[] = {
+        (GCallback) optional_check_changed, 0);
+    gtk_box_pack_start ((GtkBox *) output_checks_box, output_16bit_cbtn, false,
+        false, 0);
+    gtk_box_pack_start ((GtkBox *) output_checks_box, output_stereo_cbtn, false,
+        false, 0);
+
+    emulator_changed ();
+
+    return output_checks_box;
+}
+#endif
+
+#ifdef USE_QT
+static QCheckBox * output_16bit_cbtn, * output_stereo_cbtn;
+
+void emulator_changed ()
+{
+    if (aud_get_int (CFG_ID, "Emulator") == ADPLUG_NUKED) {
+        aud_set_bool (CFG_ID, "16bit", true);
+        aud_set_bool (CFG_ID, "Stereo", true);
+        output_16bit_cbtn->setCheckState(Qt::Checked);
+        output_16bit_cbtn->setEnabled (false);
+        output_stereo_cbtn->setCheckState(Qt::Checked);
+        output_stereo_cbtn->setEnabled (false);
+    }
+    else {
+        output_16bit_cbtn->setEnabled (true);
+        output_stereo_cbtn->setEnabled (true);
+    }
+}
+
+void * create_optional_checks ()
+{
+    QWidget * output_checks_widget;
+    QVBoxLayout * output_checks_box;
+
+    output_checks_widget = new QWidget ();
+    output_checks_box = audqt::make_vbox (output_checks_widget, audqt::sizes.TwoPt);
+    output_16bit_cbtn = new QCheckBox (
+        N_("16-bit output (if unchecked, output is 8-bit)"));
+    output_16bit_cbtn->setCheckState(
+        aud_get_bool (CFG_ID, "16bit") ? Qt::Checked : Qt::Unchecked);
+    QObject::connect (output_16bit_cbtn, & QCheckBox::stateChanged, 
+        [] (int state) {
+            aud_set_bool (CFG_ID, "16bit", state != Qt::Unchecked);
+        });
+    output_stereo_cbtn = new QCheckBox (
+        N_("Duplicate mono output to two channels"));
+    output_stereo_cbtn->setCheckState(
+        aud_get_bool (CFG_ID, "Stereo") ? Qt::Checked : Qt::Unchecked);
+    QObject::connect (output_stereo_cbtn, & QCheckBox::stateChanged, 
+        [] (int state) {
+            aud_set_bool (CFG_ID, "Stereo", state != Qt::Unchecked);
+        });
+    output_checks_box->addWidget(output_16bit_cbtn);
+    output_checks_box->addWidget(output_stereo_cbtn);
+
+    emulator_changed ();
+
+    return output_checks_widget;
+}
+#endif
+
+static const ComboItem plugin_combo[] = {
+    ComboItem ("Nuked OPL3 (Nuke.YKT, 2018)", ADPLUG_NUKED),
+    ComboItem ("WoodyOPL (DOSBox, 2016)", ADPLUG_WOODY),
+    ComboItem ("Tatsuyuki Satoh 0.72 (MAME, 2003)", ADPLUG_MAME),
+    ComboItem ("Ken Silverman (2001)", ADPLUG_KS),
+};
+
 const PreferencesWidget AdPlugXMMS::widgets[] = {
     WidgetLabel (N_("<b>Output</b>")),
-    WidgetCheck (N_("16-bit output (if unchecked, output is 8-bit)"),
-        WidgetBool (CFG_ID, "16bit")),
-    WidgetCheck (N_("Duplicate mono output to two channels"),
-        WidgetBool (CFG_ID, "Stereo")),
+    WidgetCombo (N_("OPL Emulator:"),
+        WidgetInt (CFG_ID, "Emulator", emulator_changed),
+        {{plugin_combo}}),
+#ifdef USE_GTK
+    WidgetCustomGTK (create_optional_checks),
+#endif
+#ifdef USE_QT
+    WidgetCustomQt (create_optional_checks),
```

Comment:
Since these checks are harmless to display even with emulation engines that do not support them, please just declare the widgets in the toolkit-agnostic way.

Reply:
As in removing the ifdefs? That's doable. The optional checkboxes have to be done on a per-toolkit basis so that they can be hidden or shown depending on which OPL renderer is used as the Nuked renderer doesn't support 8 bit or non-stereo output.

Reply:
After rereading your comment, I now understand what you're getting at. While, yes, leaving the checkboxes doesn't hurt anything, I'm a firm proponent that a GUI should reflect underlying capabilities instead of misrepresenting them. So I'd either say that the checkboxes should be enabled or disabled according to the capabilities of the underlying plugin, or the checkboxes should just be removed as it's rather meaningless to support 8-bit and non-stereo rendering.

Reply:
Lets just remove the checkboxes for now.  If they were critical, I would suggest that we could extend the abstraction layer, but I don't think it is worth it for 8-bit audio output ;)

Reply:
> as it's rather meaningless to support 8-bit and non-stereo rendering.

For Mono/Stereo, it depends on various factors: (1) how it's done by the internal audio mixer, (2) whether the emulated audio hardware did have two output channels, (3) whether original users connected the audio hardware to basic speakers built into monitors, and (4) whether many of the original users deliberately decided to ignore hardware output restrictions and used audio devices that mixed everything to Mono output for an overall better listening experience. For example, if using headphones, to hear the drums/percussions channel only on the left and other instruments on the right side wasn't anything like pretty, so many people preferred Mono output.

Typically, 8-bit output in emulators is available as an option for users of 8-bit audio hardware, which should not be an issue anymore, and for purists, who claim that they hear the differences in audio converted to 16-bit or higher precision.

Reply:
In reality, the 8-bit output is pointless anyway because we scale everything up to 16-bit minimum.

Reply:
Dropped the 8-bit output support as it is entirely pointless, as the core will change it to floating point anyway.

Reply:
Removed checkboxes in 8bedf01.